File: Algorithm_Mentoring/25-1/python/6주차/가장_긴_바이토닉_부분_수열.py
import sys
sys.setrecursionlimit(10**7)
input = sys.stdin.readline

def solve(cur, cnt):
    if dp[cur][cnt] != -1:
        return dp[cur][cnt]
    
    ret = 1
    # 감소하는 상태
    if cnt == 1:
        for i in range(1, cur):
            # 이전 값이 현재 값보다 작으면 건너 뜀
            if a[i] <= a[cur]:
                continue
            ret = max(ret, solve(i, 0) + 1, solve(i, 1) + 1)
    # 증가하는 상태
    else:
        for i in range(1, cur):
            # 이전 값이 현재 값보다 크면 건너 뜀
            if a[i] >= a[cur]:
                continue
            ret = max(ret, solve(i, 0) + 1)

    dp[cur][cnt] = ret
    return ret

n = int(input())
A = list(map(int, input().split()))
a = [0] + A

dp = [[-1] * 2 for _ in range(n + 1)]

ans = 0
for i in range(1, n + 1):
    # DP[i][0], DP[i][1]중 최댓값이 정답
    ans = max(ans, solve(i, 0), solve(i, 1))

print(ans)

File: Algorithm_Mentoring/25-2/cpp/8주차/가장_긴_바이토닉_부분_수열.cpp
#include <iostream>
using namespace std;
using ll = long long;

const ll MAX = 1010;
ll dp[MAX][2], a[MAX], n;

ll solve(ll cur, ll cnt){
    ll& ret = dp[cur][cnt];
    if(ret != -1) return ret; ret = 1;

    // 감소하는 상태
    if(cnt) for(int i = 1;i < cur;i++){
        // 이전 값이 현재 값보다 작으면 건너 뜀
        if(a[i] <= a[cur]) continue;
        ret = max(ret, solve(i, 0) + 1);
        ret = max(ret, solve(i, 1) + 1);
    }
    // 증가하는 상태
    else for(int i = 1;i < cur;i++){
        // 이전 값이 현재 값보다 크면 건너 뜀
        if(a[i] >= a[cur]) continue;
        ret = max(ret, solve(i, 0) + 1);
    }

    return ret;
}

int main(){
    ios::sync_with_stdio(0); // fastio
    cin.tie(0), cout.tie(0); // fastio

    cin >> n;
    for(int i = 1;i <= n;i++) cin >> a[i];
    for(int i = 0;i < MAX;i++){
        for(int j = 0;j <= 1;j++) dp[i][j] = -1;
    }

    ll result = 0;
    for(int i = 1;i <= n;i++){
        // DP[i][0], DP[i][1] 중 최댓값이 정답
        result = max(result, max(solve(i, 0), solve(i, 1)));
    }
    cout << result;
    return 0;
}

